File: internal/policy/routing.go
// agency-gateway/internal/policy/routing.go
package policy

import (
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"time"

	"gopkg.in/yaml.v3"

	"github.com/geoffbelknap/agency/internal/models"
)

// ParameterDomains maps policy parameter names to their default exception domain.
var ParameterDomains = map[string]string{
	"risk_tolerance":                 "security",
	"max_concurrent_tasks":           "operations",
	"max_task_duration":              "operations",
	"autonomous_interrupt_threshold": "security",
	"network_mediation":              "security",
	"logging":                        "compliance",
	"constraints_readonly":           "security",
	"llm_credentials_isolated":       "security",
}

// ExceptionRequest is a pending exception request awaiting routing and approval.
type ExceptionRequest struct {
	RequestID           string                   `yaml:"request_id"`
	AgentName           string                   `yaml:"agent_name"`
	Parameter           string                   `yaml:"parameter"`
	RequestedValue      string                   `yaml:"requested_value"`
	Reason              string                   `yaml:"reason"`
	Domain              string                   `yaml:"domain"`
	Status              string                   `yaml:"status"` // pending, routed, approved, denied
	RoutedTo            []string                 `yaml:"routed_to"`
	Approvals           []map[string]interface{} `yaml:"approvals"`
	Denials             []map[string]interface{} `yaml:"denials"`
	Recommendations     []map[string]interface{} `yaml:"recommendations"`
	RequiresDualApproval bool                    `yaml:"requires_dual_approval"`
	CreatedAt           string                   `yaml:"created_at"`
	ResolvedAt          string                   `yaml:"resolved_at,omitempty"`
}

// NewExceptionRequest creates a new ExceptionRequest with defaults applied.
func NewExceptionRequest(requestID, agentName, parameter, requestedValue, reason, domain string) *ExceptionRequest {
	if domain == "" {
		if d, ok := ParameterDomains[parameter]; ok {
			domain = d
		} else {
			domain = "general"
		}
	}
	return &ExceptionRequest{
		RequestID:            requestID,
		AgentName:            agentName,
		Parameter:            parameter,
		RequestedValue:       requestedValue,
		Reason:               reason,
		Domain:               domain,
		Status:               "pending",
		RoutedTo:             []string{},
		Approvals:            []map[string]interface{}{},
		Denials:              []map[string]interface{}{},
		Recommendations:      []map[string]interface{}{},
		RequiresDualApproval: false,
		CreatedAt:            time.Now().UTC().Format("2006-01-02T15:04:05Z"),
	}
}

// ToMap serialises the request to a plain map for YAML output.
func (r *ExceptionRequest) ToMap() map[string]interface{} {
	m := map[string]interface{}{
		"request_id":             r.RequestID,
		"agent_name":             r.AgentName,
		"parameter":              r.Parameter,
		"requested_value":        r.RequestedValue,
		"reason":                 r.Reason,
		"domain":                 r.Domain,
		"status":                 r.Status,
		"routed_to":              r.RoutedTo,
		"approvals":              r.Approvals,
		"denials":                r.Denials,
		"recommendations":        r.Recommendations,
		"requires_dual_approval": r.RequiresDualApproval,
		"created_at":             r.CreatedAt,
	}
	if r.ResolvedAt != "" {
		m["resolved_at"] = r.ResolvedAt
	}
	return m
}

// FromMap deserialises an ExceptionRequest from a plain map (loaded from YAML).
func FromMap(data map[string]interface{}) (*ExceptionRequest, error) {
	requestID, _ := data["request_id"].(string)
	agentName, _ := data["agent_name"].(string)
	parameter, _ := data["parameter"].(string)
	requestedValue := fmt.Sprintf("%v", data["requested_value"])
	reason, _ := data["reason"].(string)
	domain, _ := data["domain"].(string)

	req := NewExceptionRequest(requestID, agentName, parameter, requestedValue, reason, domain)

	if status, ok := data["status"].(string); ok {
		req.Status = status
	}
	// routed_to may be []string (from in-memory ToMap) or []interface{} (from YAML unmarshal).
	switch v := data["routed_to"].(type) {
	case []string:
		req.RoutedTo = append(req.RoutedTo, v...)
	case []interface{}:
		for _, item := range v {
			if s, ok := item.(string); ok {
				req.RoutedTo = append(req.RoutedTo, s)
			}
		}
	}

	// approvals
	switch v := data["approvals"].(type) {
	case []map[string]interface{}:
		req.Approvals = append(req.Approvals, v...)
	case []interface{}:
		for _, item := range v {
			if m, ok := item.(map[string]interface{}); ok {
				req.Approvals = append(req.Approvals, m)
			}
		}
	}

	// denials
	switch v := data["denials"].(type) {
	case []map[string]interface{}:
		req.Denials = append(req.Denials, v...)
	case []interface{}:
		for _, item := range v {
			if m, ok := item.(map[string]interface{}); ok {
				req.Denials = append(req.Denials, m)
			}
		}
	}

	// recommendations
	switch v := data["recommendations"].(type) {
	case []map[string]interface{}:
		req.Recommendations = append(req.Recommendations, v...)
	case []interface{}:
		for _, item := range v {
			if m, ok := item.(map[string]interface{}); ok {
				req.Recommendations = append(req.Recommendations, m)
			}
		}
	}
	if v, ok := data["requires_dual_approval"].(bool); ok {
		req.RequiresDualApproval = v
	}
	if v, ok := data["created_at"].(string); ok {
		req.CreatedAt = v
	}
	if v, ok := data["resolved_at"].(string); ok {
		req.ResolvedAt = v
	}
	return req, nil
}

// ExceptionRouter routes exception requests to appropriate principals for approval.
// Uses principals.yaml exception_routes to determine who reviews what.
// Falls back to the operator if no route is defined.
type ExceptionRouter struct {
	Home        string
	RequestsDir string
}

// NewExceptionRouter creates a new ExceptionRouter. If home is empty it defaults to ~/.agency.
func NewExceptionRouter(home string) *ExceptionRouter {
	if home == "" {
		h, err := os.UserHomeDir()
		if err != nil {
			h = "."
		}
		home = filepath.Join(h, ".agency")
	}
	return &ExceptionRouter{
		Home:        home,
		RequestsDir: filepath.Join(home, "exception-requests"),
	}
}

// loadPrincipals reads principals.yaml from home. Returns nil if missing or unreadable.
func (er *ExceptionRouter) loadPrincipals() *models.PrincipalsConfig {
	path := filepath.Join(er.Home, "principals.yaml")
	data, err := os.ReadFile(path)
	if err != nil {
		return nil
	}
	var cfg models.PrincipalsConfig
	if err := yaml.Unmarshal(data, &cfg); err != nil {
		return nil
	}
	_ = cfg.Validate()
	return &cfg
}

// Route routes an exception request to the appropriate approvers.
// Looks up the domain in principals.yaml exception_routes.
// Falls back to operator if no route is defined.
func (er *ExceptionRouter) Route(request *ExceptionRequest) (*ExceptionRequest, error) {
	principals := er.loadPrincipals()

	if principals != nil {
		// Check explicit exception_routes first.
		for _, route := range principals.ExceptionRoutes {
			if route.Domain == request.Domain {
				request.RoutedTo = make([]string, len(route.Approvers))
				copy(request.RoutedTo, route.Approvers)
				request.RequiresDualApproval = route.RequiresDualApproval
				request.Status = "routed"
				if err := er.save(request); err != nil {
					return nil, err
				}
				return request, nil
			}
		}

		// Check humans with matching exception_domains.
		var domainPrincipals []string
		for _, h := range principals.Humans {
			if h.Status == "active" {
				for _, d := range h.ExceptionDomains {
					if d == request.Domain {
						domainPrincipals = append(domainPrincipals, h.ID)
						break
					}
				}
			}
		}
		if len(domainPrincipals) > 0 {
			request.RoutedTo = domainPrincipals
			request.Status = "routed"
			if err := er.save(request); err != nil {
				return nil, err
			}
			return request, nil
		}
	}

	// Fallback: route to operator.
	request.RoutedTo = []string{"operator"}
	request.Status = "routed"
	if err := er.save(request); err != nil {
		return nil, err
	}
	return request, nil
}

// Approve records an approval from a principal.
// If dual approval is required, the request stays routed until two approvals are received.
func (er *ExceptionRouter) Approve(requestID, principalID string) (*ExceptionRequest, error) {
	request, err := er.Get(requestID)
	if err != nil {
		return nil, err
	}
	if request == nil {
		return nil, nil
	}

	if !contains(request.RoutedTo, principalID) {
		return nil, nil
	}

	request.Approvals = append(request.Approvals, map[string]interface{}{
		"principal_id": principalID,
		"timestamp":    time.Now().UTC().Format("2006-01-02T15:04:05Z"),
	})

	if request.RequiresDualApproval {
		if len(request.Approvals) >= 2 {
			request.Status = "approved"
			request.ResolvedAt = time.Now().UTC().Format("2006-01-02T15:04:05Z")
		}
	} else {
		request.Status = "approved"
		request.ResolvedAt = time.Now().UTC().Format("2006-01-02T15:04:05Z")
	}

	if err := er.save(request); err != nil {
		return nil, err
	}
	return request, nil
}

// Deny denies an exception request.
func (er *ExceptionRouter) Deny(requestID, principalID, reason string) (*ExceptionRequest, error) {
	request, err := er.Get(requestID)
	if err != nil {
		return nil, err
	}
	if request == nil {
		return nil, nil
	}

	request.Denials = append(request.Denials, map[string]interface{}{
		"principal_id": principalID,
		"reason":       reason,
		"timestamp":    time.Now().UTC().Format("2006-01-02T15:04:05Z"),
	})
	request.Status = "denied"
	request.ResolvedAt = time.Now().UTC().Format("2006-01-02T15:04:05Z")

	if err := er.save(request); err != nil {
		return nil, err
	}
	return request, nil
}

// Recommend records a function agent's recommendation on an exception request.
// Recommendations are advisory — they don't change request status.
// action must be "approve" or "deny".
// The recommender is validated by checking that agents/{agentName}/agent.yaml exists.
func (er *ExceptionRouter) Recommend(requestID, agentName, action, reasoning string) (*ExceptionRequest, error) {
	if action != "approve" && action != "deny" {
		return nil, nil
	}

	request, err := er.Get(requestID)
	if err != nil {
		return nil, err
	}
	if request == nil {
		return nil, nil
	}

	// Validate recommender: check that the agent exists.
	if !er.validateRecommender(agentName) {
		return nil, nil
	}

	request.Recommendations = append(request.Recommendations, map[string]interface{}{
		"agent":     agentName,
		"action":    action,
		"reasoning": reasoning,
		"timestamp": time.Now().UTC().Format("2006-01-02T15:04:05Z"),
	})

	if err := er.save(request); err != nil {
		return nil, err
	}
	return request, nil
}

// validateRecommender checks if the named agent exists by looking for agents/{name}/agent.yaml.
func (er *ExceptionRouter) validateRecommender(agentName string) bool {
	agentFile := filepath.Join(er.Home, "agents", agentName, "agent.yaml")
	_, err := os.Stat(agentFile)
	return err == nil
}

// Get loads an exception request by ID.
func (er *ExceptionRouter) Get(requestID string) (*ExceptionRequest, error) {
	path := filepath.Join(er.RequestsDir, requestID+".yaml")
	data, err := os.ReadFile(path)
	if os.IsNotExist(err) {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}
	var raw map[string]interface{}
	if err := yaml.Unmarshal(data, &raw); err != nil {
		return nil, err
	}
	return FromMap(raw)
}

// ListPending lists all pending/routed exception requests.
func (er *ExceptionRouter) ListPending() ([]*ExceptionRequest, error) {
	return er.listFiltered(func(req *ExceptionRequest) bool {
		return req.Status == "pending" || req.Status == "routed"
	})
}

// ListForPrincipal lists exception requests routed to a specific principal.
func (er *ExceptionRouter) ListForPrincipal(principalID string) ([]*ExceptionRequest, error) {
	return er.listFiltered(func(req *ExceptionRequest) bool {
		return req.Status == "routed" && contains(req.RoutedTo, principalID)
	})
}

func (er *ExceptionRouter) listFiltered(keep func(*ExceptionRequest) bool) ([]*ExceptionRequest, error) {
	entries, err := os.ReadDir(er.RequestsDir)
	if os.IsNotExist(err) {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}

	// Collect and sort filenames for deterministic ordering.
	var names []string
	for _, e := range entries {
		if !e.IsDir() && filepath.Ext(e.Name()) == ".yaml" {
			names = append(names, e.Name())
		}
	}
	sort.Strings(names)

	var results []*ExceptionRequest
	for _, name := range names {
		data, err := os.ReadFile(filepath.Join(er.RequestsDir, name))
		if err != nil {
			continue
		}
		var raw map[string]interface{}
		if err := yaml.Unmarshal(data, &raw); err != nil {
			continue
		}
		req, err := FromMap(raw)
		if err != nil {
			continue
		}
		if keep(req) {
			results = append(results, req)
		}
	}
	return results, nil
}

func (er *ExceptionRouter) save(request *ExceptionRequest) error {
	if err := os.MkdirAll(er.RequestsDir, 0o755); err != nil {
		return err
	}
	data, err := yaml.Marshal(request.ToMap())
	if err != nil {
		return err
	}
	path := filepath.Join(er.RequestsDir, request.RequestID+".yaml")
	return os.WriteFile(path, data, 0o644)
}

// contains reports whether s is in the slice.
func contains(slice []string, s string) bool {
	for _, v := range slice {
		if v == s {
			return true
		}
	}
	return false
}

File: images/tests/support/policy/routing.py
"""Exception routing — routes two-key exceptions to appropriate principals.

When an agent needs an exception to a policy parameter, the exception
request is routed to the principals responsible for that domain.
Privacy exceptions go to the privacy officer, security exceptions to
the security lead, etc.

Route definitions live in principals.yaml under exception_routes.
"""

import logging
from datetime import datetime, timezone
from pathlib import Path

import yaml

from typing import Optional
from images.models.principal import PrincipalsConfig

log = logging.getLogger(__name__)


# Default domain mappings for parameters when no explicit route exists
_PARAMETER_DOMAINS = {
    "risk_tolerance": "security",
    "max_concurrent_tasks": "operations",
    "max_task_duration": "operations",
    "autonomous_interrupt_threshold": "security",
    "network_mediation": "security",
    "logging": "compliance",
    "constraints_readonly": "security",
    "llm_credentials_isolated": "security",
}


class ExceptionRequest:
    """A pending exception request awaiting routing and approval."""

    def __init__(
        self,
        request_id: str,
        agent_name: str,
        parameter: str,
        requested_value: str,
        reason: str,
        domain: Optional[str] = None,
    ):
        self.request_id = request_id
        self.agent_name = agent_name
        self.parameter = parameter
        self.requested_value = requested_value
        self.reason = reason
        self.domain = domain or _PARAMETER_DOMAINS.get(parameter, "general")
        self.status = "pending"  # pending, routed, approved, denied
        self.routed_to: list[str] = []
        self.approvals: list[dict] = []  # {principal_id, timestamp}
        self.denials: list[dict] = []
        self.recommendations: list[dict] = []  # {agent, action, reasoning, timestamp}
        self.requires_dual_approval = False
        self.created_at = datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")
        self.resolved_at: Optional[str] = None

    def to_dict(self) -> dict:
        return {
            "request_id": self.request_id,
            "agent_name": self.agent_name,
            "parameter": self.parameter,
            "requested_value": self.requested_value,
            "reason": self.reason,
            "domain": self.domain,
            "status": self.status,
            "routed_to": self.routed_to,
            "approvals": self.approvals,
            "denials": self.denials,
            "recommendations": self.recommendations,
            "requires_dual_approval": self.requires_dual_approval,
            "created_at": self.created_at,
            "resolved_at": self.resolved_at,
        }

    @classmethod
    def from_dict(cls, data: dict) -> "ExceptionRequest":
        req = cls(
            request_id=data["request_id"],
            agent_name=data["agent_name"],
            parameter=data["parameter"],
            requested_value=str(data["requested_value"]),
            reason=data.get("reason", ""),
            domain=data.get("domain"),
        )
        req.status = data.get("status", "pending")
        req.routed_to = data.get("routed_to", [])
        req.approvals = data.get("approvals", [])
        req.denials = data.get("denials", [])
        req.recommendations = data.get("recommendations", [])
        req.requires_dual_approval = data.get("requires_dual_approval", False)
        req.created_at = data.get("created_at", "")
        req.resolved_at = data.get("resolved_at")
        return req


class ExceptionRouter:
    """Routes exception requests to appropriate principals for approval.

    Uses principals.yaml exception_routes to determine who reviews what.
    Falls back to the operator if no route is defined.
    """

    def __init__(self, agency_home: Optional[Path] = None):
        self.home = agency_home or Path.home() / ".agency"
        self.requests_dir = self.home / "exception-requests"

    def _load_principals(self) -> Optional[PrincipalsConfig]:
        principals_file = self.home / "principals.yaml"
        if not principals_file.exists():
            return None
        try:
            data = yaml.safe_load(principals_file.read_text())
            return PrincipalsConfig.model_validate(data)
        except Exception as e:
            log.warning("Failed to load principals: %s", e)
            return None

    def route(self, request: ExceptionRequest) -> ExceptionRequest:
        """Route an exception request to the appropriate approvers.

        Looks up the domain in principals.yaml exception_routes.
        Falls back to operator if no route is defined.
        """
        principals = self._load_principals()

        if principals:
            for route in principals.exception_routes:
                if route.domain == request.domain:
                    request.routed_to = list(route.approvers)
                    request.requires_dual_approval = route.requires_dual_approval
                    request.status = "routed"
                    self._save(request)
                    return request

            # Check if any human has this domain in exception_domains
            domain_principals = [
                h.id for h in principals.humans
                if request.domain in h.exception_domains
                and h.status == "active"
            ]
            if domain_principals:
                request.routed_to = domain_principals
                request.status = "routed"
                self._save(request)
                return request

        # Fallback: route to operator
        request.routed_to = ["operator"]
        request.status = "routed"
        self._save(request)
        return request

    def approve(
        self, request_id: str, principal_id: str
    ) -> Optional[ExceptionRequest]:
        """Record an approval from a principal.

        If dual approval is required, the request stays routed until
        two approvals are received.
        """
        request = self.get(request_id)
        if not request:
            return None

        if principal_id not in request.routed_to:
            return None

        request.approvals.append({
            "principal_id": principal_id,
            "timestamp": datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ"),
        })

        if request.requires_dual_approval:
            if len(request.approvals) >= 2:
                request.status = "approved"
                request.resolved_at = datetime.now(timezone.utc).strftime(
                    "%Y-%m-%dT%H:%M:%SZ"
                )
        else:
            request.status = "approved"
            request.resolved_at = datetime.now(timezone.utc).strftime(
                "%Y-%m-%dT%H:%M:%SZ"
            )

        self._save(request)
        return request

    def deny(
        self, request_id: str, principal_id: str, reason: str = ""
    ) -> Optional[ExceptionRequest]:
        """Deny an exception request."""
        request = self.get(request_id)
        if not request:
            return None

        request.denials.append({
            "principal_id": principal_id,
            "reason": reason,
            "timestamp": datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ"),
        })
        request.status = "denied"
        request.resolved_at = datetime.now(timezone.utc).strftime(
            "%Y-%m-%dT%H:%M:%SZ"
        )
        self._save(request)
        return request

    def recommend(
        self,
        request_id: str,
        agent_name: str,
        action: str,
        reasoning: str,
    ) -> Optional[ExceptionRequest]:
        """Record a function agent's recommendation on an exception request.

        Function agents can recommend 'approve' or 'deny' with reasoning.
        Recommendations are advisory — they don't change the request status.
        Human approvers see them when reviewing.
        """
        if action not in ("approve", "deny"):
            return None

        request = self.get(request_id)
        if not request:
            return None

        # Validate the recommender is a function agent in a relevant team
        valid = self._validate_recommender(agent_name, request.domain)
        if not valid:
            return None

        request.recommendations.append({
            "agent": agent_name,
            "action": action,
            "reasoning": reasoning,
            "timestamp": datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ"),
        })

        self._save(request)
        return request

    def _validate_recommender(self, agent_name: str, domain: str) -> bool:
        """Check if an agent is a function agent that can recommend for this domain."""
        try:
            return False  # coordination module removed; function agent validation unavailable
            team = find_agent_team(agent_name, agency_home=self.home)
            if not team:
                return False
            member = team.get_member(agent_name)
            if not member or member.agent_type != "function":
                return False
            return True
        except Exception:
            return False

    def get(self, request_id: str) -> Optional[ExceptionRequest]:
        """Load an exception request by ID."""
        req_file = self.requests_dir / f"{request_id}.yaml"
        if not req_file.exists():
            return None
        try:
            data = yaml.safe_load(req_file.read_text())
            return ExceptionRequest.from_dict(data)
        except Exception:
            return None

    def list_pending(self) -> list[ExceptionRequest]:
        """List all pending/routed exception requests."""
        if not self.requests_dir.exists():
            return []
        results = []
        for f in sorted(self.requests_dir.glob("*.yaml")):
            try:
                data = yaml.safe_load(f.read_text())
                req = ExceptionRequest.from_dict(data)
                if req.status in ("pending", "routed"):
                    results.append(req)
            except Exception:
                continue
        return results

    def list_for_principal(self, principal_id: str) -> list[ExceptionRequest]:
        """List exception requests routed to a specific principal."""
        if not self.requests_dir.exists():
            return []
        results = []
        for f in sorted(self.requests_dir.glob("*.yaml")):
            try:
                data = yaml.safe_load(f.read_text())
                req = ExceptionRequest.from_dict(data)
                if principal_id in req.routed_to and req.status == "routed":
                    results.append(req)
            except Exception:
                continue
        return results

    def _save(self, request: ExceptionRequest) -> None:
        self.requests_dir.mkdir(parents=True, exist_ok=True)
        (self.requests_dir / f"{request.request_id}.yaml").write_text(
            yaml.dump(request.to_dict(), default_flow_style=False)
        )
